flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java

```diff
@@ -816,32 +833,115 @@ public String asSummaryString() {
                     }
                     boolean isRetained =
                             FilterUtils.isRetainedAfterApplyingFilterPredicates(
                                     filterPredicates, getValueGetter(row));
                     if (isRetained) {
                         final Row projectedRow = projectRow(row);
-                        final RowData rowData = (RowData) converter.toInternal(projectedRow);
-                        if (rowData != null) {
-                            if (numRetained >= numElementToSkip) {
-                                rowData.setRowKind(row.getKind());
-                                result.add(rowData);
-                            }
-                            numRetained++;
-                        }
+                        resultBuffer.add(projectedRow);
+                    }
+                }
+                if (overLimit) {
+                    break;
+                }
+            }
+            // simulate aggregate operation
+            if (!aggregateExpressions.isEmpty()) {
+                resultBuffer = applyAggregatesToRows(resultBuffer);
+            }
+            List<RowData> result = new ArrayList<>();
+            for (Row row : resultBuffer) {
+                final RowData rowData = (RowData) converter.toInternal(row);
+                if (rowData != null) {
+                    if (numRetained >= numElementToSkip) {
+                        rowData.setRowKind(row.getKind());
+                        result.add(rowData);
+                    }
+                    numRetained++;
+                }
+            }
+            return result;
+        }
+
+        private List<Row> applyAggregatesToRows(List<Row> rows) {
+            if (groupingSet != null && groupingSet.length > 0) {
+                // has group by, group firstly
+                Map<Row, List<Row>> buffer = new HashMap<>();
+                for (Row row : rows) {
+                    Row bufferKey = new Row(groupingSet.length);
+                    for (int i = 0; i < groupingSet.length; i++) {
+                        bufferKey.setField(i, row.getField(groupingSet[i]));
+                    }
+                    if (buffer.containsKey(bufferKey)) {
+                        buffer.get(bufferKey).add(row);
+                    } else {
+                        buffer.put(bufferKey, new ArrayList<>(Collections.singletonList(row)));
                     }
                 }
+                List<Row> result = new ArrayList<>();
+                for (Map.Entry<Row, List<Row>> entry : buffer.entrySet()) {
+                    result.add(Row.join(entry.getKey(), accumulateRows(entry.getValue())));
+                }
+                return result;
+            } else {
+                return Collections.singletonList(accumulateRows(rows));
+            }
+        }
+
+        // can only apply sum/sum0/avg function for long type fields for testing
+        private Row accumulateRows(List<Row> rows) {
```

Comment:
add limitation for`applyAggregates` method which can only support long aggregations ?